File: dotnet/base/Mcma.Core/Logging/LogEvent.cs
using System;
using Mcma.Serialization;
using Newtonsoft.Json.Linq;

namespace Mcma.Logging
{
    public class LogEvent
    {
        public LogEvent(
            string type,
            int level,
            string source,
            string requestId,
            DateTime timestamp, 
            string message,
            object[] args,
            McmaTracker tracker = null)
        {
            Type = type;
            Level = level;
            Source = source;
            RequestId = requestId;
            Timestamp = timestamp;
            Message = message;
            Args = args;
            Tracker = tracker;
        }
        
        public string Type { get; }
        public int Level  { get; }
        public string Source  { get; }
        public string RequestId { get; }
        public DateTimeOffset Timestamp  { get; }
        public string Message { get; }
        public object[] Args { get; }
        public McmaTracker Tracker { get; }

        public override string ToString()
        {
            var json = (JObject)McmaJson.ToMcmaJson(this);

            var trackerProperty = json.Property(nameof(Tracker), StringComparison.OrdinalIgnoreCase);
            if (trackerProperty != null)
                trackerProperty.Remove();

            if (Tracker != null)
            {
                json["trackerId"] = Tracker.Id;
                json["trackerLabel"] = Tracker.Label;

                if (Tracker.Custom != null)
                    foreach (var customProperty in Tracker.Custom)
                    {
                        var customPropertyKey =
                            nameof(Tracker.Id).Equals(customProperty.Key, StringComparison.OrdinalIgnoreCase) || 
                            nameof(Tracker.Label).Equals(customProperty.Key, StringComparison.OrdinalIgnoreCase)
                                ? $"trackerCustom{customProperty.Key}"
                                : $"tracker{customProperty.Key}";

                        json[customPropertyKey] = customProperty.Value;
                    }
            }

            return json.ToString();
        }
    }
}

File: base/core/lib/logging/log-event.ts
import * as util from "util";
import { McmaTrackerProperties } from "../model";
import { McmaException } from "../mcma-exception";

export class LogEvent {

    constructor(public type: string,
                public level: number,
                public source: string,
                public requestId: string,
                public timestamp: Date,
                public message: any,
                public tracker?: McmaTrackerProperties) {
    }

    flatten(): { [key: string]: any } {
        let message = this.message;
        if (message instanceof Error) {
            if (this.message.stack) {
                message = this.message.stack;
            } else {
                message = this.message.toString();
            }
        }
        const logEventEntry: { [key: string]: any } = {
            type: this.type,
            level: this.level,
            source: this.source,
            requestId: this.requestId,
            timestamp: this.timestamp,
            message,
            trackerId: this.tracker?.id,
            trackerLabel: this.tracker?.label,
        };

        if (this.tracker?.custom) {
            for (const key of Object.keys(this.tracker?.custom)) {
                const trackerCustomKey = key.substring(0, 1).toUpperCase() + key.substring(1);
                const trackerCustomValue = this.tracker?.custom[key];

                switch (trackerCustomKey) {
                    case "Id":
                    case "Label":
                        logEventEntry["trackerCustom" + trackerCustomKey] = trackerCustomValue;
                        break;
                    default:
                        logEventEntry["tracker" + trackerCustomKey] = trackerCustomValue;
                        break;
                }
            }
        }

        return logEventEntry;
    }

    toString(): string {
        const logEventEntry = this.flatten();
        try {
            return JSON.stringify(logEventEntry, null, 2);
        } catch {
            try {
                logEventEntry.message = util.inspect(logEventEntry.message);
            } catch {
                logEventEntry.message = logEventEntry.message + "";
            }
        }

        return JSON.stringify(logEventEntry, null, 2);
    }
}
